### PR TITLE
[Feature] add auto_step flag in WandbVisBackend

### PR DESCRIPTION
**Pull request by i-aki-y**
## Motivation

The wandb backend uses a mechanism of the auto-increment built in the `wandb.log`.
But this step information will be reset to zero after the resumption.
So when I resume the training, I get the following discontinuous graph.

resume_auto_step

This behavior is a little confusing, I think the graphs should be continuous before and after the resumption.

resume_manual_step

## Modification

To fix this, I added an `auto_step` flag that allows the user to select the increment method.
- auto_step == True: Keep the current behavior (Default)
- auto_step == False: Use the step given by an argument. Note that other loggers also use this argument as the step.

## BC-breaking (Optional)

No. Default auto_step is set to True which is consistent with the current behavior.

## Use cases (Optional)

This makes wandb graph natural before and after the resumption (like the second image above).

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/visualization/vis_backend.py [hunk 1]
```diff
@@ -371,6 +371,8 @@ class WandbVisBackend(BaseVisBackend):
         auto_step: (bool) Whether use auto step increment built-in
             ``wandb.log`` or not. The built-in step will be reset to zero when
             the training is resumed. Default: True
+            Note that if you set False, you also should set
+            ``log_metric_by_epoch=False`` in the LoggerHook.
```
Comment (HAOCHENYE) [inline mmengine/visualization/vis_backend.py:375]:
```suggestion

    Note:
        If ``auto_step`` is False, ``log_metric_by_epoch`` in :cls:`mmengine.hook.LoggerHook` needs to be set to ``False`` too.
```

File: mmengine/visualization/vis_backend.py [hunk 2]
```diff
@@ -368,6 +368,11 @@ class WandbVisBackend(BaseVisBackend):
             New in version 0.3.0.
         watch_kwargs (optional, dict): Agurments for ``wandb.watch``.
             New in version 0.4.0.
+        auto_step: (bool) Whether use auto step increment built-in
+            ``wandb.log`` or not. The built-in step will be reset to zero when
+            the training is resumed. Default: True
```
Comment (HAOCHENYE) [inline mmengine/visualization/vis_backend.py:373]:
```suggestion
        auto_step: (bool) Whether use auto step increment built-in
            ``wandb.log`` or not. The built-in step will be reset to zero when
            the training is resumed. Default: True
            New in version 0.7.3
```

**Review by HAOCHENYE — COMMENTED** [review]
Thanks for your contribution. The overall modification LGTM :smile: .

**Comment by i-aki-y** [comment]
@HAOCHENYE  Thank you for your review.

File: mmengine/visualization/vis_backend.py [hunk 3]
```diff
@@ -368,6 +368,13 @@ class WandbVisBackend(BaseVisBackend):
...
+            New in version 0.7.3
+
+    Note:
+        If ``auto_step`` is False, ``log_metric_by_epoch`` in :cls:`mmengine.hook.LoggerHook` needs to be set to ``False`` too.
```
Comment (HAOCHENYE) [inline mmengine/visualization/vis_backend.py:377]: Needs a line break here.

File: mmengine/visualization/vis_backend.py [hunk 4]
```diff
@@ -465,10 +478,13 @@ def add_scalar(self,
...
-            step (int): Useless parameter. Wandb does not
-                need this parameter. Defaults to 0.
+            step (int): Global step value to record. Defaults to 0.
+                This will be ignored if auto_step is True.
         """
-        self._wandb.log({name: value}, commit=self._commit)
+        if self._auto_step:
+            self._wandb.log({name: value}, commit=self._commit)
+        else:
+            self._wandb.log({name: value}, commit=self._commit, step=step)
```
[the remark this answers is not in the document]
Comment (i-aki-y) [reply mmengine/visualization/vis_backend.py:487]: @zhouzaida  thank you for your review.

To make it work, we need to change the default value of the `log_metric_by_epoch` in the `LoggerHook` as described in the "Note" section in this docstring.
I am not sure I can change the default value like `log_metric_by_epoch` in this PR.

The followings are the reason why `log_metric_by_epoch=False` is needed.

First, the `wandb.log` only supports monotonically increasing step sequences.
(See, https://docs.wandb.ai/guides/track/log/logging-faqs#why-am-i-seeing-fewer-data-points-than-i-logged )
If we set an earlier step in `wandb.log`, the `wandb` looks to ignore the log.
In the `LoggerHook`, the `runner.iter` is used as the step, but there is an exception, the `after_val_epoch` uses `runner.epoch` by default. 
Since almost always the `runner.epoch < runner.iter` is satisfied, we need to set `log_metric_by_epoch=False` to avoid the `wandb` ignoring the validation metrics.


One idea to remove the `auto_step` flag is to add a new attribute, `last_step`, in the `WandbVisBackend`. 
When the step < self.last_step is satisfied (this is probably runner.epoch), replace it with the `last_step`.

```py
if step < self.last_step:
    step = self.last_step
else:
    self.last_step = step
self._wandb.log(..., step=step)
```

I think this can ensure the step is always monotonically increasing. This may be more robust and better than `auto_flag` approach because we do not need to care about the value of `log_metric_by_epoch`.
Should I try this version?

**Comment by i-aki-y** [comment]
Sorry, I am reconsidering this PR because my first version is not user-friendly a little. In my first version, the `auto_step` interacts with `LoggerHook`'s `log_metric_by_epoch`, but I think this interaction is non-trivial for many users. And if users misconfigure them, the `wandb` will ignore the validation metrics without any errors. This misconfiguration could often happen in a complicated system like mmdet and confuse users.

File: mmengine/visualization/vis_backend.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhouzaida) [reply mmengine/visualization/vis_backend.py:487]: Hi @i-aki-y , thanks for your very clear explanation and I got your point. Due to the feature of wandb.log, we should increase the `step` monotonically. The idea you proposed looks good to me.

**Comment by zhouzaida** [comment]
> Sorry, I am reconsidering this PR because my first version is not user-friendly. In my first version, the `auto_step` interacts with `LoggerHook`'s `log_metric_by_epoch`, but I think this interaction is non-trivial for many users. And if users misconfigure them, the `wandb` will ignore the validation metrics without any errors. This misconfiguration could often happen in a complicated system like mmdet and confuse users.

Hi @i-aki-y , the idea your proposed above looks good to me.

**Comment by i-aki-y** [comment]
@zhouzaida @HAOCHENYE I revised the PR, I would appreciate it if you review it.

As I answered @zhouzaida's comment, I introduced the `self._last_step` (`+ 1` is needed) to avoid an invalid step being passed to `wandb.log`. As a result, the implementation becomes a bit complicated, but I think this is worth doing. Now I could remove the note for the `log_metric_by_epoch`.

I still left the `auto_step` flag to keep the previous behavior. But switching the `auto_flag` changes only the unit of step (from the "count of `wandb.log` calls" to the `runner.iter`) and resumption behavior.
If such a change in the step's unit is acceptable, let me know, and I will remove the `auto_flag` to always use `wandb.log(...,step=step)`.

**Comment by HAOCHENYE** [comment]
Hi! The modification makes sense to me. Refers to https://github.com/wandb/wandb/issues/2642 and https://docs.wandb.ai/ref/python/run#define_metric, I found that WandB can support logging data on different x axis like that:

```python
import wandb

# 1. Start a W&B run
wandb.init(project='test')

# 2. Save model inputs and hyperparameters
config = wandb.config
config.learning_rate = 0.01
wandb.define_metric('loss1', step_metric='iter', step_sync=False)
wandb.define_metric('loss2', step_metric='epoch', step_sync=False)
# Model training code here ...

# 3. Log metrics over time to visualize performance
for i in range(100):
    wandb.log({"loss1": i, 'iter': i})
    if i % 10 == 0:
        wandb.log({"loss2": i, 'epoch': i // 10})
        wandb.log({"loss3": i * 10, 'epoch': i // 10})
```
Will this make it possible to record epoch-based data?

**Comment by i-aki-y** [comment]
@HAOCHENYE  Thank you for pointing interesting feature.

It looks to work as expected!

Unfortunately, the current `define_metric_cfg` option of the `WandbVisBackend` only supports a case of `define_metric(metric, summary=summary)`, but if we can fix this and makes include the "iter" (and “epoch" when by_epoch==True), to the `tag` of the log output, we can control the behavior of the wandb plot more flexibly.

**Comment by HAOCHENYE** [comment]
> Author

I think this kind of modification `define_metric_cfg={'coco/bbox_mAP': 'max', step_type: 'epoch'}` is OK

**Comment by i-aki-y** [comment]
@HAOCHENYE  thank you for the advice.

I think we should allow setting multiple define_metrics in the cfg.
Can I re-define the existing `define_metric_cfg` as a list of dicts from a dict, or should I introduce a new argument like `define_metric_cfg"s”`?

I think the following usage:

```
vis_backends = [
    dict(type='LocalVisBackend'),
    dict(
        type='WandbVisBackend',        
        init_kwargs={...},
        define_metric_cfgs=[
            dict(name="lr", step_metric='iter’),
            ...
            dict(name="coco/bbox_mAP", step_metric='epoch'),
            dict(name="coco/bbox_mAP_50", step_metric='epoch'),            
        ],
    )
]
```

**Comment by i-aki-y** [comment]
@HAOCHENYE @zhouzaida  Since the solution that uses define_metric is based on a different idea and the implementation is also different,  I made a separate PR (#1099).

If the new solution is better I will close this issue.

**Comment by zhouzaida** [comment]
> @HAOCHENYE thank you for the advice.
> 
> I think we should allow setting multiple define_metrics in the cfg. Can I re-define the existing `define_metric_cfg` as a list of dicts from a dict, or should I introduce a new argument like `define_metric_cfg"s”`?
> 
> I think the following usage:
> 
> ```
> vis_backends = [
>     dict(type='LocalVisBackend'),
>     dict(
>         type='WandbVisBackend',        
>         init_kwargs={...},
>         define_metric_cfgs=[
>             dict(name="lr", step_metric='iter’),
>             ...
>             dict(name="coco/bbox_mAP", step_metric='epoch'),
>             dict(name="coco/bbox_mAP_50", step_metric='epoch'),            
>         ],
>     )
> ]
> ```

Hi, we can re-define the existing `define_metric_cfg` which can be a dict or list of dict.

**Comment by zhouzaida** [comment]
> @HAOCHENYE @zhouzaida Since the solution that uses define_metric is based on a different idea and the implementation is also different, I made a separate PR (#1099).
> 
> If the new solution is better I will close this issue.

The solution in #1099 maybe a more elegant solution so we can close this PR. Thanks for your contribution very much.

**Comment by i-aki-y** [comment]
I think PR (https://github.com/open-mmlab/mmengine/pull/1099) is a better solution, so I am closing this PR
